src/hotspot/share/oops/method.cpp

```diff
@@ -238,27 +243,73 @@ int Method::fast_exception_handler_bci_for(const methodHandle& mh, Klass* ex_kla
     int beg_bci = table.start_pc(i);
     int end_bci = table.end_pc(i);
     assert(beg_bci <= end_bci, "inconsistent exception table");
+    log_debug(exceptions)("  - checking exception table entry for BCI %d to %d",
+                         beg_bci, end_bci);
+
     if (beg_bci <= throw_bci && throw_bci < end_bci) {
       // exception handler bci range covers throw_bci => investigate further
+      log_debug(exceptions)("    - entry covers throw point BCI %d", throw_bci);
+
       int handler_bci = table.handler_pc(i);
       int klass_index = table.catch_type_index(i);
       if (klass_index == 0) {
+        if (log_is_enabled(Debug, exceptions)) {
+          log_debug(exceptions)("    - found catch-all handler at BCI: %d", handler_bci);
+        } else if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("Found catch-all handler for exception of type \"%s\" in method \"%s\" at BCI: %d",
+                               ex_klass == NULL ? "NULL" : ex_klass->external_name(), mh()->name()->as_C_string(), handler_bci);
+        }
         return handler_bci;
       } else if (ex_klass == NULL) {
+        // Is this even possible?
+        if (log_is_enabled(Debug, exceptions)) {
+          log_debug(exceptions)("    - NULL exception class is implicitly caught by handler at BCI: %d", handler_bci);
+        } else if (log_is_enabled(Info, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_info(exceptions)("NULL exception class is implicitly caught by handler in method \"%s\" at BCI: %d",
+                               mh()->name()->as_C_string(), handler_bci);
+        }
         return handler_bci;
       } else {
+        if (log_is_enabled(Debug, exceptions)) {
+          ResourceMark rm(THREAD);
+          log_debug(exceptions)("    - resolving catch type \"%s\"",
+                               pool->klass_name_at(klass_index)->as_C_string());
+        }
         // we know the exception class => get the constraint class
         // this may require loading of the constraint class; if verification
         // fails or some other exception occurs, return handler_bci
-        Klass* k = pool->klass_at(klass_index, CHECK_(handler_bci));
+        Klass* k = pool->klass_at(klass_index, THREAD);
+        if (HAS_PENDING_EXCEPTION) {
+        if (log_is_enabled(Debug, exceptions)) {
```

Comment:
Technically, this would be interesting enough to be log_info level.

Reply:
Perhaps, but again making it fit and provide sufficient contextual information is awkward and the throw is already captured at the Info level. So I would hope that if someone saw the unexpected throw logging they would use that as a cue to turn on debug logging to get more information.

Logging is really, really subjective - painfully so.

Reply:
ok.